### PR TITLE
feat: ability to change name in profile

### PR DESCRIPTION
**Pull request by Sboonny**
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

This aim to allow user to update their name in their profile (`userProfile`)

### CODE REVIEW

**Review by gikf — COMMENTED** [review]
The _original_ name in form doesn't change after saving, this makes reverting the change require reloading page:
- Change name from A to B.
- Press save button. Name B is saved.
- Change name in input to the name A.
- Save button is disabled.

I cannot make suggestions in these places in github, this can be fixed with following:
```
--- a/client/src/modules/profiles/component/ProfileForm.tsx
+++ b/client/src/modules/profiles/component/ProfileForm.tsx
@@ -1,5 +1,5 @@
 import { Button } from '@chakra-ui/react';
-import React from 'react';
+import React, { useEffect } from 'react';
 import { useForm } from 'react-hook-form';
 import { Input } from '../../../components/Form/Input';
 import { TextArea } from '../../../components/Form/TextArea';
@@ -40,11 +40,16 @@ export const ProfileForm: React.FC<ProfileFormProps> = (props) => {
   const {
     handleSubmit,
     register,
+    reset,
     formState: { isDirty },
   } = useForm<UpdateUserInputs>({
     defaultValues,
   });
 
+  useEffect(() => {
+    reset({ name: me });
+  }, [me]);
+
   return (
     <Form submitLabel={submitText} FormHandling={handleSubmit(onSubmit)}>
       {fields.map(({ key, label, placeholder, required, type }) =>
```

---

Another very specific issue is that client and server will accept blank spaces as name - ` `.

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 1]
```diff
@@ -48,6 +73,14 @@ export const UserProfilePage = () => {
             </>
           )}
 
+          <ProfileForm
+            loading={loadingUpdate}
+            onSubmit={submitUpdateMe}
+            data={user.name}
+            loadingText={'Saving Chapter Changes'}
+            submitText={'Save Chapter Changes'}
```
Comment (gikf) [inline client/src/modules/profiles/pages/userProfile.tsx:81]:
```suggestion
            loadingText={'Saving Profile Changes'}
            submitText={'Save Profile Changes'}
```

**Comment by ojeytonwilliams** [comment]
> Another very specific issue is that client and server will accept blank spaces as name - .

I think an empty string is fine, but we should trim them so we don't have ' ' and '   ' be different names.

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 2]
```diff
@@ -15,6 +22,24 @@ export const UserProfilePage = () => {
...
+    refetchQueries: [{ query: meQuery }],
+  });
+
+  const submitUpdateMe = async (data: UpdateUserInputs) => {
+    setLoadingUpdate(true);
+    try {
+      await updateMe({
+        variables: { data },
+      });
+      await router.push('/profile');
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/pages/userProfile.tsx:35]:
```suggestion
```
Is this necessary?

File: client/src/modules/profiles/component/ProfileForm.tsx [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
 import { TextArea } from '../../../components/Form/TextArea';
 import { Form } from '../../../components/Form/Form';
-import { MeQuery, UpdateUserInputs } from '../../../generated/graphql';
+import { UpdateUserInputs } from '../../../generated/graphql';
 
 interface ProfileFormProps {
   loading: boolean;
   onSubmit: (data: UpdateUserInputs) => Promise<void>;
-  data: MeQuery;
+  data: string;
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/component/ProfileForm.tsx:12]:
```suggestion
  data: UpdateUserInputs;
```

File: client/src/modules/profiles/component/ProfileForm.tsx [hunk 4]
```diff
@@ -35,18 +35,21 @@ const fields: Fields[] = [
 export const ProfileForm: React.FC<ProfileFormProps> = (props) => {
   const { loading, onSubmit, data, submitText, loadingText } = props;
 
-  const me = data?.me;
-  const defaultValues: UpdateUserInputs = {
-    name: me?.name || '',
-  };
+  const me = data;
+  const defaultValues: UpdateUserInputs = { name: me };
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/component/ProfileForm.tsx:39]:
```suggestion
  const defaultValues: UpdateUserInputs = { name: data.name };
```

File: client/src/modules/profiles/component/ProfileForm.tsx [hunk 5]
```diff
@@ -35,18 +35,21 @@ const fields: Fields[] = [
...
     register,
+    reset,
     formState: { isDirty },
   } = useForm<UpdateUserInputs>({
     defaultValues,
   });
 
+  useEffect(() => {
+    reset({ name: me });
+  }, [me]);
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/component/ProfileForm.tsx:51]:
```suggestion
    reset({ name: data.name });
  }, [data]);
```

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 6]
```diff
@@ -48,6 +73,14 @@ export const UserProfilePage = () => {
             </>
           )}
 
+          <ProfileForm
+            loading={loadingUpdate}
+            onSubmit={submitUpdateMe}
+            data={user.name.trim()}
```
Comment (ojeytonwilliams) [inline client/src/modules/profiles/pages/userProfile.tsx:79]:
```suggestion
            data={user}
```
This is more type-safe.  We have to provide `name`, not just any string.  We can trim the data in `submitUpdateMe`

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/modules/profiles/pages/userProfile.tsx [hunk 7]
```diff
@@ -48,6 +72,14 @@ export const UserProfilePage = () => {
             </>
           )}
 
+          <ProfileForm
+            loading={loadingUpdate}
+            onSubmit={submitUpdateMe}
+            data={user}
```
Comment (Sboonny) [inline client/src/modules/profiles/pages/userProfile.tsx:78]:
```suggestion
            data={data}
```

**Review by Sboonny — COMMENTED** [review]
Reapply the suggestion

**Review by ojeytonwilliams — APPROVED** [review]
Thanks @Sboonny, this LGTM :+1:

**Review by ojeytonwilliams — APPROVED** [review]
I finally remembered we supposed to trim :cold_sweat: 

LGTM :+1:

**Pull request merged**